### PR TITLE
ARROW-15134: [GLib] Add GArrow{Month,DayTime,MonthDayNano}IntervalDataType

### PR DESCRIPTION
**Pull request by okadakk**

### CODE REVIEW

File: c_glib/arrow-glib/basic-data-type.cpp [hunk 1]
```diff
@@ -97,6 +97,15 @@ G_BEGIN_DECLS
  * #GArrowTime64DataType is a class for the number of microseconds or
  * nanoseconds since midnight in the 64-bit signed integer data type.
  *
+ * #GArrowIntervalDataType is an abstract class for interval related data type
+ * such as #GArrowIntervalMonthsDataType.
+ *
+ * #GArrowIntervalMonthsDataType is a class for the month intarval data type.
```
Comment (kou) [inline c_glib/arrow-glib/basic-data-type.cpp:103]: We should use `GArrowXXXIntervalDataType` for consistency.
For example, we already have `GArrowListDataType` and `GArrowLargeListDataType`.

For `arrow::IntervalType::MONTHS`, let's use `MONTH` not `MONTHS` in GLib for consistency. Other interval types don't have `S`.

We can discuss C++ API (`arrow::IntervalType::MONTHS` or `...::MONTH`, which is better?) later with C++ developers on Jira or mailing list.

File: c_glib/arrow-glib/basic-data-type.cpp [hunk 2]
```diff
@@ -1287,6 +1296,137 @@ garrow_time64_data_type_new(GArrowTimeUnit unit, GError **error)
...
+static void
+garrow_interval_data_type_class_init(GArrowIntervalDataTypeClass *klass)
+{
+}
+
+/**
+ * garrow_interval_data_type_get_interval_type:
+ * @type: #GArrowIntervalDataType.
+ *
+ * Returns: GArrowIntervalType
```
Comment (kou) [inline c_glib/arrow-glib/basic-data-type.cpp:1317]: We need more description here.

File: c_glib/arrow-glib/basic-data-type.cpp [hunk 3]
```diff
@@ -1287,6 +1296,137 @@ garrow_time64_data_type_new(GArrowTimeUnit unit, GError **error)
...
+ * Returns: The newly created interval months data type.
+ *
+ * Since: 7.0.0
+ */
+GArrowIntervalMonthsDataType *
+garrow_interval_months_data_type_new(void)
+{
+  auto arrow_data_type = arrow::month_interval();
+
+  GArrowIntervalMonthsDataType *data_type =
```
Comment (kou) [inline c_glib/arrow-glib/basic-data-type.cpp:1356]: We can use `auto` here.

File: c_glib/arrow-glib/basic-data-type.cpp [hunk 4]
```diff
@@ -1287,6 +1296,137 @@ garrow_time64_data_type_new(GArrowTimeUnit unit, GError **error)
...
+/**
+ * garrow_interval_data_type_get_interval_type:
+ * @type: #GArrowIntervalDataType.
+ *
+ * Returns: GArrowIntervalType
+ *
+ * Since: 7.0.0
+ */
+GArrowIntervalType
+garrow_interval_data_type_get_interval_type(GArrowIntervalDataType *type) {
```
Comment (kou) [inline c_glib/arrow-glib/basic-data-type.cpp:1322]: We use the following style:

```c++
TYPE
FUNCTION_NAME(ARGS)
{
```

File: c_glib/arrow-glib/basic-data-type.h [hunk 5]
```diff
@@ -457,6 +457,65 @@ GArrowTime64DataType *garrow_time64_data_type_new      (GArrowTimeUnit unit,
...
+                         garrow_interval_data_type,
+                         GARROW,
+                         INTERVAL_DATA_TYPE,
+                         GArrowTimeDataType)
+struct _GArrowIntervalDataTypeClass
+{
+  GArrowTimeDataTypeClass parent_class;
+};
+
+GArrowIntervalType garrow_interval_data_type_get_interval_type(GArrowIntervalDataType *type);
```
Comment (kou) [inline c_glib/arrow-glib/basic-data-type.h:471]: We need to add `GARROW_AVAILABLE_IN_7_0` annotation for new functions.

File: c_glib/arrow-glib/type.h [hunk 6]
```diff
@@ -94,6 +95,7 @@ typedef enum {
   GARROW_TYPE_TIME64,
   GARROW_TYPE_INTERVAL_MONTHS,
   GARROW_TYPE_INTERVAL_DAY_TIME,
+  GARROW_TYPE_INTERVAL_MONTH_DAY_NANO,
```
Comment (kou) [inline c_glib/arrow-glib/type.h:98]: We should not put this here.
We need to use the same order as `arrow::Type::type` to use the same integer value.

File: c_glib/arrow-glib/type.h [hunk 7]
```diff
@@ -126,4 +128,18 @@ typedef enum {
...
 } GArrowTimeUnit;
 
+/**
+ * GArrowIntervalType:
+ * @GARROW_INTERVAL_TYPE_MONTHS: Months.
+ * @GARROW_INTERVAL_TYPE_DAY_TIME: Days milliseconds.
+ * @GARROW_INTERVAL_TYPE_MONTH_DAY_NANO: Months days nanoseconds.
+ *
+ * They are corresponding to `arrow::IntervalType::type` values.
+ */
```
Comment (kou) [inline c_glib/arrow-glib/type.h:138]: `Since: 7.0.0` is missing.

File: c_glib/test/test-interval-day-time-data-type.rb [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+class TestIntervalDayTimeDataType < Test::Unit::TestCase
+  def test_type
+    data_type = Arrow::IntervalDayTimeDataType.new
```
Comment (kou) [inline c_glib/test/test-interval-day-time-data-type.rb:20]: We can use `setup` for common setup code:

```ruby
def setup
  @data_type = Arrow::IntervalDayTimeDataType.new
end

def test_type
  assert_equal(..., @data_type.id)
end

def test_xxx
  assert_equal(..., @data_type.xxx)
end
```

**Review by kou — COMMENTED** [review]
Thanks.

I'll push fixes of my comments soon. Please confirm them.

**Comment by okadakk** [comment]
@kou 
I learned a lot!
Thank you for your review!

**Comment by kou** [comment]
Thanks for confirming this.
I'll merge this once CI is passed.

**Review by kou — APPROVED** [review]
+1

**Pull request closed without merging**